Ready-to-test: Sync nginx config with .htaccess

RealRancor opened the pull request:
Ready to test

Fixes #1297

Xenopathic commented on admin_manual/installation/nginx_configuration.rst:
```diff
@@ -73,11 +73,23 @@ Nginx Configuration
      rewrite ^/.well-known/host-meta.json /public.php?service=host-meta-json last;
 
+     rewrite ^/apps/calendar/caldav.php /remote.php/caldav/ last;
+     rewrite ^/apps/contacts/carddav.php /remote.php/carddav/ last;
```
Personally, I think these two rewrites should be removed from the `.htaccess`, not added here

RealRancor replied:
I thought about the same but if we can't get a statement that those are not needed anymore (https://github.com/owncloud/documentation/issues/1297#issuecomment-119694919) i think its better safe than sorry.

josh4trunks commented:
I don't see a reason for /apps/X/Ydav.php to redirect to /remote.php/Ydav unless we know of any apps that actually use the /apps URI.

/.wellknown URIs are defined here so applications may be using them https://tools.ietf.org/html/rfc6764
We should keep these, but possibly use return instead of rewrite?

RealRancor commented:
@josh4trunks 
> I don't see a reason for /apps/X/Ydav.php to redirect to /remote.php/Ydav unless we know of any apps that actually use the /apps URI.

Those are old/legacy rules from pre-4.0.x times:

https://github.com/owncloud/core/issues/17651#issuecomment-122079511

Seems they can die according to @LukasReschke in the comment above.

> /.wellknown URIs are defined here so applications may be using them https://tools.ietf.org/html/rfc6764
We should keep these, but possibly use return instead of rewrite?

I don't know the difference here but can have a look and do some further tests.

RealRancor commented:
Updated PR, ready to test.

RealRancor commented on admin_manual/installation/nginx_configuration.rst:
```diff
@@ -71,7 +67,11 @@ Nginx Configuration
+      deny all;
+    }
+
+    location ~* ^/(\.user.ini|\.htaccess|\.mailmap|\.tag|autotest|occ|issue|indie|db_|console).* {
```
I can't use 

``location ~* ^/(\.|autotest|occ|issue|indie|db_|console).* {``

here as this is also blocking the access to the .well-known URLs.

RealRancor commented on admin_manual/installation/nginx_configuration.rst:
```diff
@@ -80,8 +80,10 @@ Nginx Configuration
       rewrite ^/.well-known/host-meta /public.php?service=host-meta last;
       rewrite ^/.well-known/host-meta.json /public.php?service=host-meta-json last;
```
Those two seems to causing a 404 / Not Found here and also on the demo.owncloud.org page:

```
curl -i https://demo.owncloud.org/.well-known/host-meta

HTTP/1.1 404 Not Found
Server: nginx
Date: Tue, 13 Oct 2015 12:36:42 GMT
Content-Type: text/html; charset=UTF-8
Content-Length: 0
Connection: keep-alive
Set-Cookie: 52400eaba833a=n8ce4tpqm2877082fcp1u5r2i5; path=/; HttpOnly
Expires: Thu, 19 Nov 1981 08:52:00 GMT
Cache-Control: no-store, no-cache, must-revalidate, post-check=0, pre-check=0
Pragma: no-cache
Content-Security-Policy: default-src 'self'; script-src stats.owncloud.org 'self' 'unsafe-eval' 'unsafe-inline'; style-src 'self' 'unsafe-inline'; frame-src *; img-src *; font-src 'self' data:; media-src *; connect-src *
X-XSS-Protection: 1; mode=block
X-Content-Type-Options: nosniff
X-Frame-Options: Sameorigin
X-Robots-Tag: none
Accept-Ranges: bytes
Age: 0
Via: seetheprogress caching
X-Powered-By: community guy
X-Paid: No
```

RealRancor replied:
Mhhh, a local Apache Installation is also doing the same. Thats one big mess :-/

```
curl -i http://127.0.0.1/.well-known/host-meta.json

HTTP/1.0 404 Not Found
Date: Tue, 13 Oct 2015 14:22:32 GMT
Server: Apache
X-Powered-By: PHP/5.5.9-1ubuntu4.13
Set-Cookie: ocsz3lwy561l=4mo8jbiki1lrbgnqcip3504d37; path=/; HttpOnly
Expires: Thu, 19 Nov 1981 08:52:00 GMT
Cache-Control: no-store, no-cache, must-revalidate, post-check=0, pre-check=0
Pragma: no-cache
Content-Security-Policy: default-src 'self'; script-src 'self' 'unsafe-eval'; style-src 'self' 'unsafe-inline'; frame-src *; img-src *; font-src 'self' data:; media-src *; connect-src *
X-XSS-Protection: 1; mode=block
X-Content-Type-Options: nosniff
X-Frame-Options: Sameorigin
X-Robots-Tag: none
Content-Length: 0
Connection: close
Content-Type: text/html; charset=UTF-8
```

RealRancor replied:
Ok, nevermind. This is only needed for the user_webfinger app (https://github.com/owncloud/apps/tree/master/user_webfinger)

RealRancor commented on admin_manual/installation/nginx_configuration.rst:
```diff
@@ -80,8 +80,10 @@ Nginx Configuration
+      rewrite ^/.well-known/carddav /remote.php/carddav/ permanent;
+      rewrite ^/.well-known/caldav /remote.php/caldav/ permanent;
+
+      rewrite ^/remote/(.*) /remote.php last;
```
This is also something i really don't know if it its needed. Just tested and on Apache as well as on nginx a 404 is caused by this:

```
 curl -i http://127.0.0.1/remote/abc
HTTP/1.1 404 Not Found
Date: Tue, 13 Oct 2015 12:43:31 GMT
Server: Apache
Content-Length: 267
Content-Type: text/html; charset=iso-8859-1

<!DOCTYPE HTML PUBLIC "-//IETF//DTD HTML 2.0//EN">
<html><head>
<title>404 Not Found</title>
</head><body>
<h1>Not Found</h1>
<p>The requested URL /remote/abc was not found on this server.</p>
<hr>
<address>Apache Server at 127.0.0.1 Port 80</address>
</body></html>
```

josh4trunks commented:
thanks for doing all this exploration. the config is gonna be a lot cleaner/sane because of your efforts!

as for those deny all regex. I don't see a reason we need to match all cases "~*" as we know what those files names are. I assume the "~" is way faster/less work because it doesn't need to look for every possible UpPeR/lOwEr combination.

RealRancor commented:
Thanks again for the feedback. Just have updated the config. Frank didn't like the idea to remote the legacy rewrites from the main .htaccess so i'm just added them here.

RealRancor commented:
Also just removed the ``gzip off;`` based on the info provided in https://github.com/owncloud/client/issues/3946#issuecomment-147985427

oC 8.1 introduced the OC-ETag header which is not changed when gzipping the content.

RealRancor commented:
Sorry, two new updates:

1. Removed the legacy rules again based on the discussion in https://github.com/owncloud/core/pull/19802 Those where never working for Apache so no reason to add this again to the nginx config

2. Added ``gzip off;`` until it is sorted out that all apps are also using the new OC-ETag header (https://github.com/owncloud/client/issues/3946#issuecomment-148342764)

carlaschroder commented:
@RealRancor @josh4trunks Thanks for all your work on this. Is it ready to merge?

josh4trunks commented:
:thumbsup: thanks @RealRancor for all the effort

RealRancor commented:
@carlaschroder Ready to merge from my PoV.

What i have tested:

1. Service discovery from DAVDroid via the .well-known URLs
2. Correct blocking of the resources within the deny all blocks

I wasn't able to test ``rewrite ^/remote/(.*) /remote.php last;`` as even in core no one knows the purpose of this rule.

josh4trunks commented:
Hey @RealRancor I noticed something while looking at the deny regex today.
Is there a reason we add the wildcard match ``.*`` these regex should match regardless and deny access to the resources before any other location matches. In my opinion this should be removed

Also, I assume admins never want to serve "hidden" files starting with "."
Maybe we should start the match with ``(?:\.|`` which would simplify the config and match all "." files. Might be a good idea to implement this in the htaccess as well.

RealRancor commented:
@josh4trunks 
> Is there a reason we add the wildcard match .* at the end?
Thats the same syntax used in the core .htaccess. I would like to keep both in sync for easier maintenance.

> Maybe we should start the match with (?:\.|

At my tests this caused an access forbidden to the .well-known urls for service discovery. Thats why i have only added the current existing hidden files from core.

josh4trunks commented:
ok, not sure how htaccess matches work but for nginx regex, as long as you don't end with $ it implies anything can go after. So ending with ``.*`` is redundant. If this isn't needed for htaccess then I recommend we remove it in htaccess and in nginx

ahh, didn't think of the .well-known URLs! Did you try putting those redirects before the deny location? regex's match in top down order. I suggest we take them out of the / location because, as I believe you stated the other day, .well-known need to be at the root regardless of owncloud being hosted in a subdirectory. I suggest these live next to the robots.txt location, above the deny location.

josh4trunks commented:
We could even make it so everything that is relative of a subdirectory/is portable, is nested in the ``/`` location. that way it's easy to modify the config for hosting owncloud in a subdirectory.
The only stuff that would need to live outside is robots.txt and the well-know redirects.

carlaschroder commented:
Suggestion: Mark all these Nginx configs as WIP until they're actually ready to merge :)

RealRancor commented:
@josh4trunks 
> ok, not sure how htaccess matches work but for nginx regex, as long as you don't end with $ it implies anything can go after. So ending with .* is redundant. If this isn't needed for htaccess then I recommend we remove it in htaccess and in nginx

Ah, good to know. Will update the config accordingly in the next few days.

> ahh, didn't think of the .well-known URLs! Did you try putting those redirects before the deny location? regex's match in top down order. I suggest we take them out of the / location because, as I believe you stated the other day, .well-known need to be at the root regardless of owncloud being hosted in a subdirectory. I suggest these live next to the robots.txt location, above the deny location.

Sounds good. Will test that also.

@carlaschroder 
> Suggestion: Mark all these Nginx configs as WIP until they're actually ready to merge :)

It marked as WIP until it was "ready-to test" https://github.com/owncloud/documentation/pull/1320#issuecomment-147685948 :-)

RealRancor commented:
@josh4trunks 

Updated and already tested the PR. Works fine for me, even with the deny of all hidden files.

The only thing i have noticed is, that when accessing an known file via:

http://example.com/data/$user/files/known_file.txt

i'm getting an 404/file not found instead of the 403/forbidden. This is caused by X-Sendfile and throws an 403/forbidden when disabling X-Sendfile.

josh4trunks commented:
looks good to me. :thumbsup:

I'm guessing you have a ^~ location which captures attempts in /data for X-Sendfile. you can remove this location since X-Sendfile was removed from ownCloud 8.2.0+

RealRancor commented:
> I'm guessing you have a ^~ location which captures attempts in /data for X-Sendfile.
Yes, and from what i have read the "internal" used there is returning 404 as a default.

> you can remove this location since X-Sendfile was removed from ownCloud 8.2.0+

Still running oC 8.1.3 here :-)

RealRancor commented:
@FelixDefrance As you did https://github.com/owncloud/documentation/pull/1809 could you also those here? Thanks

FelixDefrance commented:
@RealRancor as i told to @carlaschroder i'll can check it next week ;)

RealRancor commented:
Sounds great, thanks!

carlaschroder commented:
Nothing inspires feedback like publishing this stuff. So, tally ho!

RealRancor commented:
Thanks for merging. Running this since a longer time without any issues.

Backports for stable8.1 and 8.2 are in #1944 and #1945. Don't want to backport below as i havn't tested this with older versions.
